### PR TITLE
More flexible incremental fetching

### LINKED ISSUE #87: More flexible incremental fetching
ref: https://github.com/keboola/internal/issues/59

### PR DESCRIPTION
**Pull request by pivnicek**
Fixes #87

### CODE REVIEW

**Comment by pivnicek** [comment]
TODO: Incr fetching tests for `decimal` and `smalldatetime`

**Comment by pivnicek** [comment]
Live config test failing due to #91

File: tests/Keboola/DbExtractor/IncrementalFetchingTest.php [hunk 1]
```diff
@@ -70,7 +70,7 @@ public function testIncrementalFetchingByAutoIncrement(): void
         sleep(2);
         // the next fetch should be empty
```
Comment (tomasfejfar) [inline tests/Keboola/DbExtractor/IncrementalFetchingTest.php]: update comment and rname $emptyResult which is not empty

File: tests/Keboola/DbExtractor/IncrementalFetchingTest.php [hunk 2]
```diff
@@ -29,9 +29,9 @@ public function testIncrementalFetchingByTimestamp(): void
         $this->assertNotEmpty($result['state']['lastFetchedRow']);
         @unlink($outputFile);
         sleep(2);
-        // the next fetch should be empty
+        // the next fetch should be just the last fetched row from last time because of >=
         $emptyResult = ($this->createApplication($config, $result['state']))->run();
```
Comment (tomasfejfar) [inline tests/Keboola/DbExtractor/IncrementalFetchingTest.php]: $emptyResult is not empty

File: tests/Keboola/DbExtractor/IncrementalFetchingTest.php [hunk 3]
```diff
@@ -81,7 +81,79 @@ public function testIncrementalFetchingByAutoIncrement(): void
...
+            ],
+            $result['imported']
+        );
+        //check that output state contains expected information
+        $this->assertArrayHasKey('state', $result);
+        $this->assertArrayHasKey('lastFetchedRow', $result['state']);
+        $this->assertEquals(6.6, $result['state']['lastFetchedRow']);
+        unlink($outputFile);
+        sleep(2);
+        // the next fetch should be empty
```
Comment (tomasfejfar) [inline tests/Keboola/DbExtractor/IncrementalFetchingTest.php]: it's not empty ;)

File: tests/Keboola/DbExtractor/MSSQLTest.php [hunk 4]
```diff
@@ -929,7 +1063,7 @@ public function testGetTables(): void
         $this->assertArrayHasKey('tables', $result);
         $this->assertEquals('success', $result['status']);
         $this->assertCount(4, $result['tables']);
-
+
```
Comment (tomasfejfar) [inline tests/Keboola/DbExtractor/MSSQLTest.php]: You don't have https://intellij-support.jetbrains.com/hc/en-us/community/posts/205803079-Configure-editor-to-remove-whitespaces-in-blank-lines set up?

Comment (pivnicek) [reply tests/Keboola/DbExtractor/MSSQLTest.php]: ah, it was set to do it only for `modified lines`.  So it should have done it here...

File: src/Keboola/DbExtractor/Extractor/MSSQL.php [hunk 5]
```diff
@@ -12,8 +12,8 @@
 
 class MSSQL extends Extractor
 {
-    public const TYPE_AUTO_INCREMENT = 'autoIncrement';
-    public const TYPE_TIMESTAMP = 'timestamp';
+    public const INCREMENT_TYPE_NUMERIC = 'numeric';
+    public const INCREMENT_TYPE_TIMESTAMP = 'timestamp';
```
[the remark this answers is not in the document]
Comment (pivnicek) [reply src/Keboola/DbExtractor/Extractor/MSSQL.php]: I'd like to keep it consistent among extractors.  mssql is the only one with the `timestamp` naming issue.

**Comment by pivnicek** [comment]
ok @tomasfejfar I've updated this as per your suggestions, thank you.

**Comment by pivnicek** [comment]
Yes, it's not actually about an empty csv, it's an SQL server 2008 issue that I'm working on now (next PR coming today which should knock off another couple issues)

**Comment by pivnicek** [comment]
I'm going to merge this and make sure the build is fixed in the follow-up, thanks tom

**Comment by pivnicek** [comment]
Yup, it'll be fixed today.  Unfortunately the only was to test against sql server 2008 is on that live config since we can't put it in a container.

**Comment by pivnicek** [comment]
I merged it mainly to make a cleaner diff in #97

**Comment by pivnicek** [comment]
@tomasfejfar rest assured I've removed the problematic timestamp column from the sql2008 server test table so the build passes now. https://travis-ci.com/keboola/db-extractor-mssql/builds/94263562 Going to release this, and then finish that follow up.

**Pull request merged**